flow/connectors/postgres/client.go

```diff
@@ -643,35 +644,73 @@ func (c *PostgresConnector) getTableNametoUnchangedCols(
 	return resultMap, nil
 }
 
-func (c *PostgresConnector) getCurrentLSN(ctx context.Context) (pglogrepl.LSN, error) {
-	tx, err := c.conn.BeginTx(ctx, pgx.TxOptions{IsoLevel: pgx.ReadCommitted})
+func (c *PostgresConnector) executeReadOnlyTxn(ctx context.Context, fn func(tx pgx.Tx) error) error {
+	tx, err := c.conn.BeginTx(ctx, pgx.TxOptions{IsoLevel: pgx.ReadCommitted, AccessMode: pgx.ReadOnly})
 	if err != nil {
-		return 0, fmt.Errorf("begin transaction failed: %w", err)
+		return fmt.Errorf("failed to begin transaction: %w", err)
 	}
+
 	defer func() {
 		if rollbackErr := tx.Rollback(ctx); rollbackErr != nil && rollbackErr != pgx.ErrTxClosed {
 			c.logger.Error("rollback failed", slog.Any("error", rollbackErr))
 		}
 	}()
 
-	query := `
-		SELECT CASE
-			WHEN pg_is_in_recovery() THEN pg_last_wal_receive_lsn()
-			ELSE pg_current_wal_lsn()
-		END`
-
-	var lsnString string
-	if err := tx.QueryRow(ctx, query).Scan(&lsnString); err != nil {
-		return 0, fmt.Errorf("query failed: %w", err)
+	if err := fn(tx); err != nil {
```

Comment:
I don't think this accomplishes what you want still, you need to reconnect on net errors